examples/gno.land/p/demo/timelock/timelock.gno

```diff
@@ -0,0 +1,218 @@
+	timestamps    *avl.Tree
+	accessControl *accesscontrol.Role
+	minDelay      uint64
+}
+
+// New instance of TimeLockUtil
+func NewTimeLockUtil(timestamps *avl.Tree, accessControl *accesscontrol.Role, minDelay uint64) (*TimeLockUtil, error) {
+	if timestamps == nil || accessControl == nil {
+		return nil, errors.New("timestamps and accesscontrol values must be different from nil")
+	}
+
+	return &TimeLockUtil{
+		timestamps:    timestamps,
+		accessControl: accessControl,
+		minDelay:      minDelay,
+	}, nil
+}
+
+// Schedules an operation to be carried out after a minimum delay
+func (tl *TimeLockUtil) Schedule(id seqid.ID, delay uint64) error {
+	if delay < tl.minDelay {
+		return errors.New("timelockutil: Schedule: insufficient delay")
+	}
+
+	if tl.timestamps.Has(id.Binary()) {
+		return errors.New("timelockutil: Schedule: operation already scheduled")
+	}
+
+	tl.timestamps.Set(id.Binary(), uint64(time.Now().Unix())+delay)
+
+	std.Emit(
+		"TimeLockScheduled",
+		"id", id.String(),
+		"delay", strconv.FormatInt(int64(delay), 10),
+	)
+
+	return nil
+}
+
+// Remove operation
+func (tl *TimeLockUtil) Remove(id seqid.ID) {
+	tl.timestamps.Remove(id.Binary())
+
+	std.Emit(
+		"TimeLockRemoved",
+		"id", id.String(),
+	)
+}
+
+// Cancels a planned operation
+func (tl *TimeLockUtil) Cancel(id seqid.ID) error {
+	if !tl.IsPending(id) {
+		errors.New("timelock: Cancel: operation not pending")
+	}
+
+	tl.timestamps.Remove(id.Binary())
+
+	std.Emit(
+		"TimeLockCancelled",
+		"id", id.String(),
+	)
+
+	return nil
+}
+
+// Executes a pending operation
+func (tl *TimeLockUtil) Execute(id seqid.ID) error {
+	if !tl.IsPending(id) {
+		errors.New("timelock: Execute: operation not pending")
+	}
+
+	tl.timestamps.Set(id.Binary(), DoneTimestamp)
+
+	std.Emit(
+		"TimeLockExecuted",
+		"id", id.String(),
+	)
+
+	return nil
+}
+
+// Update the minimum lead time for future operations
+func (tl *TimeLockUtil) UpdateDelay(newDelay uint64) error {
+	if std.GetOrigCaller() != tl.accessControl.AdminRole {
+		return errors.New("timelock: UpdateDelay: only admin can update delay")
+	}
+
+	std.Emit(
+		"TimeLockMinDelayChanged",
+		"oldDelay", strconv.FormatInt(int64(tl.minDelay), 10),
+		"newDelay", strconv.FormatInt(int64(newDelay), 10),
+	)
+
+	tl.minDelay = newDelay
+
+	return nil
+}
+
+// Checks if an operation is pending
+func (tl *TimeLockUtil) IsPending(id seqid.ID) bool {
+	state, err := tl.GetOperationState(id)
+	if err != nil {
+		// Handle the error appropriately; for now, we assume the operation is not pending if there's an error
+		ufmt.Errorf("Error retrieving operation state: %v", err)
+		return false
+	}
+
+	return state == Pending
+}
+
+// Checks if an operation is ready
+func (tl *TimeLockUtil) IsReady(id seqid.ID) bool {
+	state, err := tl.GetOperationState(id)
+	if err != nil {
+		// Handle the error appropriately; for now, we assume the operation is not pending if there's an error
+		ufmt.Errorf("Error retrieving operation state: %v", err)
```

Comment:
does this work?

Reply:
i think you should just return `bool, error` here.

Reply:
ok i read the rest of your code, I think you should do something like this:

```go
ts := tl.GetTimestamp(id)
ts.IsReady()
```

this way you can create a `tl.MustGetTimestamp(id)` and chain it that way `tl.MustGetTimestamp(id).IsReady()`.

Reply:
Using `IsReady(id)` will produce the same result as `MustGetTimestamp(id).IsReady()`. I don't really understand what's the point of this modification; can you give me some details on how you see it, please? @moul